src/ray/util/internal/stream_redirection_utils.h

```diff
@@ -0,0 +1,45 @@
+// Copyright 2025 The Ray Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include <memory>
+
+#include "ray/util/pipe_logger.h"
+#include "ray/util/scoped_dup2_wrapper.h"
+
+namespace ray::internal {
+
+struct RedirectionHandleWrapper {
+  RedirectionFileHandle redirection_file_handle;
+  // Used for restoration.
+  std::unique_ptr<ScopedDup2Wrapper> scoped_dup2_wrapper;
+};
+
+// Util functions to redirect stdout / stderr stream based on the given redirection option
+// [opt]. This function is _NOT_ thread-safe.
+RedirectionHandleWrapper RedirectStreamImpl(MEMFD_TYPE_NON_UNIQUE stream_fd,
+                                            const StreamRedirectionOption &opt);
+
+// Synchronize on redirection handler in blocking style.
+void SyncOnStreamRedirection(RedirectionHandleWrapper &redirection_handle_wrapper);
+
+// Flush on redirected stream synchronously.
+//
+// TODO(hjiang): Current implementation is naive, which directly flushes on spdlog logger
+// and could miss those in the pipe; it's acceptable because we only use it in the unit
+// test for now.
+void FlushOnRedirectedStream(RedirectionHandleWrapper &redirection_handle_wrapper);
```

Comment:
This is only used for testing, so it isn't really part of the redirection utils interface, and it's only 1 line of code. So prefer to make it a utility inside the test file and promote it only if it is required in the real implementation.

Reply:
We have two tests using the same flush function, do you want me to copy & paste it twice?

Reply:
Link:
https://github.com/ray-project/ray/blob/cd1c2b17d7c47c3c7d9fc68497220e89a51b3c48/src/ray/util/tests/stream_redirection_exit_test.cc#L46
https://github.com/ray-project/ray/blob/cd1c2b17d7c47c3c7d9fc68497220e89a51b3c48/src/ray/util/tests/stream_redirection_utils_test.cc#L66

Reply:
actually it looks like the comment might be wrong and this is actually used in the top-level `stream_redirection_utils`? https://github.com/ray-project/ray/pull/51191/files#diff-2a7e096be7505545545d270bcc45f035b23dfa071acddaf0f67fd0a8f011d5a2R66

Reply:
All flush related functions are only used in unit tests, but not production code.